lib/extract-stream.js

```diff
@@ -9,28 +9,33 @@ function extractStream (dest, opts) {
   const sawIgnores = new Set()
   return tar.x({
     cwd: dest,
-    filter: (name, entry) => !entry.header.type.match(/^.*link$/i),
+    // Use endsWith rather than match with regex. Should be faster.
+    filter: (name, entry) => !entry.header.type.endsWith('link'),
     strip: 1,
     onwarn: msg => opts.log && opts.log.warn('tar', msg),
     uid: opts.uid,
     gid: opts.gid,
     onentry (entry) {
-      if (entry.type.toLowerCase() === 'file') {
+      // Only lowerCase once.
+      const entryType = entry.type.toLowerCase()
+      if (entryType === 'file') {
         entry.mode = opts.fmode & ~(opts.umask || 0)
-      } else if (entry.type.toLowerCase() === 'directory') {
+      } else if (entryType === 'directory') {
         entry.mode = opts.dmode & ~(opts.umask || 0)
       }
 
       // Note: This mirrors logic in the fs read operations that are
       // employed during tarball creation, in the fstream-npm module.
       // It is duplicated here to handle tarballs that are created
       // using other means, such as system tar or git archive.
-      if (entry.type.toLowerCase() === 'file') {
+      if (entryType === 'file') {
         const base = path.basename(entry.path)
         if (base === '.npmignore') {
           sawIgnores.add(entry.path)
         } else if (base === '.gitignore') {
-          const npmignore = entry.path.replace(/\.gitignore$/, '.npmignore')
+          // Cut off last 10 chars (.gitignore) and replace with .npmignore.
+          // substring should be faster than regex.
+          const npmignore = entry.path.slice(0, -10) + '.npmignore'
```

Comment:
still working on testing actual extraction, but [this gist](https://gist.github.com/diddledan/e6537530eeeababb40a4c1cfb7544b4f) shows three orders of magnitude difference between using regex with `String.prototype.replace` to using `String.prototype.slice`:

```text
------------------------------------------------
     worstcase replace regex
------------------------------------------------
  7075 ops/s @ ~0.141ms/op (±5.17%)
  Sampled 77 in 5.84s.
------------------------------------------------
     replace regex
------------------------------------------------
  20343 ops/s @ ~0.049ms/op (±5.33%)
  Sampled 78 in 5.92s.
------------------------------------------------
     substring
------------------------------------------------
  32353737 ops/s @ ~0.000ms/op (±3.68%)
  Sampled 79 in 6.05s.
------------------------------------------------
     slice
------------------------------------------------
  36347626 ops/s @ ~0.000ms/op (±4.63%) <-- my PR uses this method
  Sampled 76 in 5.97s.
------------------------------------------------
     substr
------------------------------------------------
  31818807 ops/s @ ~0.000ms/op (±4.02%)
  Sampled 78 in 6.00s.```